### PR TITLE
Task/bloated jexl 1447

### PR DESCRIPTION
**Pull request by mineralntl**
At present, the values/data of the attributes are compared directly but depending on how strict/lenient we want to be, we could do actual fuzzy string matching using something like JavaWuzzy.

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/function/JexlContextCreator.java [hunk 1]
```diff
@@ -64,7 +65,47 @@ public JexlContextCreator() {
...
-        return Tuples.tuple(from.first(), from.second(), context);
+        return Tuples.tuple(dedupedFrom.first(), dedupedFrom.second(), context);
+    }
+    
+    private Tuple3 dedupeAttributes(Tuple3<Key,Document,Map<String,Object>> rawTuple) {
+        Map<String,Attribute<? extends Comparable<?>>> dict = rawTuple.second().getDictionary();
+        Set<Attribute<? extends Comparable<?>>> dedupedAttributeSet = new HashSet<Attribute<? extends Comparable<?>>>();
+        
+        for (Entry<String,Attribute<? extends Comparable<?>>> entry : dict.entrySet()) {
+            if (entry.getKey().equals(BODY_KEY)) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/function/JexlContextCreator.java]: Why are we only doing this for the BODY_KEY?  The BODY values are in fact index only so there will be no values from the event.  Dedupping those values are reasonable but why not do that for all fields?

File: warehouse/query-core/src/main/java/datawave/query/function/JexlContextCreator.java [hunk 2]
```diff
@@ -64,7 +65,47 @@ public JexlContextCreator() {
...
+                Attributes rawButes = (Attributes) entry.getValue();
+                Set<Attribute<? extends Comparable<?>>> attributeSet = rawButes.getAttributes();
+                List<String> valueList = new ArrayList<String>();
+                
+                // get list of any non-preNormalizedAttributes since those are the ones
+                // we actually care about and want to keep
+                attributeSet.forEach(attribute -> {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/function/JexlContextCreator.java]: I think we should use the Attribute.isFromIndex() instead of checking for PreNormalizedAttribute.  That would be much less fragile.

**Review by ivakegg — CHANGES_REQUESTED** [review]

**Comment by mineralntl** [comment]
Belay this PR until further notice - spoke with @apmoriarty and need to rework the logic.

**Comment by mineralntl** [comment]
Completely changed direction based on team feedback.

File: warehouse/query-core/src/main/java/datawave/query/util/FuzzyAttributeComparator.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
+package datawave.query.util;
+
+import com.google.common.collect.Sets;
+import datawave.query.attributes.Attribute;
+import datawave.query.attributes.Attributes;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public final class FuzzyAttributeComparator {
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/util/FuzzyAttributeComparator.java]: I know I wrote the original ticket with the fuzzy wording, but let's change this class name. Fuzzy matching is a mathematical operation and in the end we're just doing a comparison of the backing data values. 

Rename to `AttributeComparator` or something similar. Add a class level comment that we're identifying attributes with equivalent data values across attribute types.

File: warehouse/query-core/src/main/java/datawave/query/util/FuzzyAttributeComparator.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    
+    // this is a utility and should never be instantiated
+    private FuzzyAttributeComparator() {
+        throw new UnsupportedOperationException();
+    }
+    
+    public static boolean singleToSingle(final Attribute existingAttribute, final Attribute newAttribute) {
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/util/FuzzyAttributeComparator.java]: Let's standardize the argument naming convention here. The method names are keeping to `single` and `multiple` but the args are `existing`, `new`, `single`, and `multiple`. You could make this a simple `left`/`right` or `leftSingle` and `rightSingle`. Or perhaps `attr` for single Attribute and `attrs` for an instance of Attributes. Just make the args consistent among themselves.

File: warehouse/query-core/src/main/java/datawave/query/util/FuzzyAttributeComparator.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        return (existingAttribute.getData().equals(newAttribute.getData())
+                        && existingAttribute.getColumnVisibility().equals(newAttribute.getColumnVisibility()) && existingAttribute.getTimestamp() == newAttribute
+                        .getTimestamp());
+    }
+    
+    public static boolean singleToMultiple(final Attributes multipleAttributes, final Attribute singleAttribute) {
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/util/FuzzyAttributeComparator.java]: Method args should follow the method name. Single attribute on the left, Multiple attributes on the right.

File: warehouse/query-core/src/main/java/datawave/query/util/FuzzyAttributeComparator.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        }
+        return containsMatch;
+    }
+    
+    /**
+     * Performs a basic fuzzy match on the content/data of two given Attributes. If such a match is found, combine the metadata from the two Attributes and
+     * create a new set of deduplicated Attributes to return to the Document.
+     *
+     * @param existingAttributes
+     *            The set of Attributes against which we want to compare
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/util/FuzzyAttributeComparator.java]: Strictly speaking, we do not have a set of attributes (`Set<Attributes>`). We have a single instance of `Attributes`, which is itself a set of type attribute (`Set<Attribute>`)

File: warehouse/query-core/src/test/java/datawave/query/util/FuzzyAttributeComparatorTest.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+import java.util.HashSet;
+import java.util.Set;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class FuzzyAttributeComparatorTest {
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/util/FuzzyAttributeComparatorTest.java]: This test looks fine overall, but let's add a `DocumentTest` and roll through a number of expected cases.
1. add two instances of `Attribute` to document
2. add two instances of `Attributes` to document
3. add a mix of `Attribute` and `Attributes` to document

For each case outlined above hit the following variants
- test case without duplicate values
- test case with duplicate value
- test case with duplicate value in different Attribute types (Content vs. PreNormalizedAttribute)

Also be sure to hit a few edge cases
- merge attribute with document key vs. no document key
- merge attribute with toKeep flag set/not set
- merge attribute with different visibilities (test will act as documentation of expected behavior)

**Review by apmoriarty — CHANGES_REQUESTED** [review]
Overall this is looking much better, would like to see a few more test cases. Give some consideration to putting this work behind a feature flag.

File: warehouse/query-core/src/main/java/datawave/query/util/FuzzyAttributeComparator.java (hunk 5 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/util/FuzzyAttributeComparator.java]: Done and done

File: warehouse/query-core/src/main/java/datawave/query/util/FuzzyAttributeComparator.java (hunk 6 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/util/FuzzyAttributeComparator.java]: Fair enough - will change the phrasing.

File: warehouse/query-core/src/main/java/datawave/query/util/FuzzyAttributeComparator.java (hunk 3 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/util/FuzzyAttributeComparator.java]: Done and done.

File: warehouse/query-core/src/main/java/datawave/query/function/JexlContextCreator.java (hunk 2 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/function/JexlContextCreator.java]: OBE

File: warehouse/query-core/src/main/java/datawave/query/function/JexlContextCreator.java (hunk 1 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/function/JexlContextCreator.java]: Updated to check all keys

File: warehouse/query-core/src/main/java/datawave/query/util/FuzzyAttributeComparator.java (hunk 4 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/util/FuzzyAttributeComparator.java]: Done and done.

File: warehouse/query-core/src/test/java/datawave/query/util/FuzzyAttributeComparatorTest.java (hunk 7 above)
Comment (mineralntl) [reply warehouse/query-core/src/test/java/datawave/query/util/FuzzyAttributeComparatorTest.java]: Updated merging logic and added new edge cases as requested

File: warehouse/query-core/src/main/java/datawave/query/attributes/Attributes.java [hunk 8]
```diff
@@ -1,5 +1,15 @@
...
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
+import datawave.marking.MarkingFunctions;
+import datawave.query.collections.FunctionalSet;
+import datawave.query.jexl.DatawaveJexlContext;
+import org.apache.commons.lang.builder.HashCodeBuilder;
+import org.apache.hadoop.io.WritableUtils;
+import org.apache.log4j.Logger;
+
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/attributes/Attributes.java]: I think you can remove this change from the set of changes....

File: warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        List<Attribute> trueDuplicates = new ArrayList<Attribute>();
+        
+        attrs1.getAttributes().forEach(attr1 -> {
+            AtomicBoolean containsMatch = new AtomicBoolean(false);
+            attrs2.getAttributes().forEach(attr2 -> {
+                if (singleToSingle(attr1, attr2) && (attr1.isToKeep() || attr2.isToKeep())) {
+                    if (attr1.getMetadata().getSize() == attr2.getMetadata().getSize()) {
+                        containsMatch.set(true);
+                        trueDuplicates.add(attr1);
+                        combinedAttrMap.put(attr1, 1); // found true match so return pre-existing Attribute
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java:93]: If !attr1.isToKeep and attr2.isToKeep, but we are only retaining attr1 then I think we need to set attr1.keep flag.

Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java:93]: same comment goes for below when retaining attr2 instead

File: warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+                containsMatch = true;
+            }
+        }
+        return containsMatch;
+    }
+    
+    /**
+     * Performs a basic match on the content/data of two given Attributes. If such a match is found, combine the metadata from the two Attributes and create new
+     * deduplicated Attributes to return to the Document.
+     *
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java:74]: document how you are choosing which attr to keep if two values are the same.

File: warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    }
+    
+    public static Set<Attribute<? extends Comparable<?>>> combineMultipleAttributes(final Attributes attrs, final Attribute attr, final boolean trackSizes) {
+        HashSet<Attribute<? extends Comparable<?>>> attrSet = Sets.newHashSet();
+        attrSet.add(attr);
+        
+        return combineMultipleAttributes(attrs, new Attributes(attrSet, attrs.isToKeep(), trackSizes));
+    }
+    
+    public static Attribute combineSingleAttributes(final Attribute attr1, final Attribute attr2) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java:149]: method comments

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        int attr1mdSize = attr1.getMetadata().getSize();
+        int attr2mdSize = attr2.getMetadata().getSize();
+        
+        Attribute mergedAttr = null;
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java:153]: We also need to look at the isFromIndex flag in here.  We need to prefer holding onto things from the event over things from the index.  This has a huge impact when using the #OCCURANCE function.

File: warehouse/query-core/src/main/java/datawave/query/attributes/Attributes.java (hunk 8 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/attributes/Attributes.java]: Noted - will undo this in the next commit

File: warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java (hunk 10 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java:74]: Will do

File: warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java (hunk 9 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java:93]: Ah good catch - will update in the next commit

File: warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java (hunk 11 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java:149]: Will add

File: warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java (hunk 12 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/util/AttributeComparator.java:153]: Added as per team request

**Comment by mineralntl** [comment]
Closing as outdated - see #1612 for updated PR
